storage/src/snapshot_cache.rs

```diff
@@ -0,0 +1,268 @@
+    /// be evicted to make space.
+    ///
+    /// [`Snapshot`]s must be inserted sequentially relative to their `jmt::Version`s and have
+    /// consecutive version numbers.
+    ///
+    /// ## Special case
+    ///
+    /// The first [`Snapshot`] inserted is allowed to have version `u64::MAX` to support the
+    /// pre-genesis case.
+    ///
+    /// ## Panics
+    ///
+    /// This function panics if the supplied snapshot's version is:
+    ///
+    /// - stale i.e. older than the most recently inserted snapshot
+    ///
+    /// - skipping a version i.e. the difference between the version numbers is greater than 1.
+    ///
+    /// For a non-faillible version, see [`try_push`].
+    pub fn push(&mut self, snapshot: Snapshot) {
+        self.try_push(snapshot)
+            .expect("failed to push snapshot in cache");
+    }
+
+    /// Returns the latest inserted `Snapshot`.
+    pub fn latest(&self) -> Snapshot {
+        self.cache
+            .front()
+            .map(Clone::clone)
+            .expect("cache cannot be empty")
+    }
+
+    /// Returns the oldest snapshot in the cache.
+    fn oldest(&self) -> Snapshot {
+        self.cache.back().unwrap().clone()
+    }
+
+    /// Attempts to fetch a [`Snapshot`] with a matching `jmt::Version`, and returns `None` if none
+    /// was found.
+    pub fn get(&self, version: jmt::Version) -> Option<Snapshot> {
+        let latest_version = self.latest().version();
+
+        // Special case: the first ever snapshot inserted is allowed to have a "pre-genesis"
+        // `jmt::Version` version (`u64::MAX`). In this case, it is guaranteed to always be
+        // at the back of the queue.
+        if version == u64::MAX {
+            let oldest_snapshot = self.oldest();
+            if version == oldest_snapshot.version() {
+                Some(oldest_snapshot)
+            } else {
+                None
+            }
+        } else if version > latest_version {
+            None
+        } else {
+            // We compute the offset assuming that snapshot entries are cached
+            // such that the delta between entries is always 1.
+            let offset = latest_version.saturating_sub(version) as usize;
+            self.cache
+                .get(offset)
+                .map(Clone::clone)
+                .filter(|s| s.version() == version)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::snapshot::Snapshot;
+    use crate::snapshot_cache::SnapshotCache;
+    use crate::storage::Storage;
+
+    fn create_storage_instance() -> Storage {
+        use tempfile::tempdir;
+        use tokio::runtime::Runtime;
+        let rt = Runtime::new().unwrap();
+        // create a storage backend for testing
+        let dir = tempdir().unwrap();
+        let file_path = dir.path().join("snapshot-cache-testing.db");
+
+        rt.block_on(Storage::load(file_path)).unwrap()
+    }
```

Comment:
The `SnapshotCache` API is synchronous so it doesn't matter in this case, but rather than manually constructing a runtime every time, we could use the `#[tokio::test]` to write the tests as `async fn`.